Readme.md
#CTA200 2022


assignment_1/history.txt
    1  1234Ruyi
    2  cosmo3.cita.utoronto.ca
    3  /Applications/TigerVNC\ Viewer\ 1.12.0.app/Contents/MacOS/TigerVNC\ Viewer 
    4  ssh student19@gw.cita.utoronto.ca
    5  ssh username@gw.cita.utoronto.ca
    6  ssh student19@gw.cita.utoronto.ca
    7  /cita/local/bin/passwd
    8      ssh student19@gw.cita.utoronto.ca
    9  cd~ murray
   10  cd ~murray
   11  ls 
   12  ls -la
   13  cd Pa/Hopkin/All
   14  cd Pa
   15  cd ~murray
   16  ls -la
   17  cd Pa
   18  cd Papers/Hopkins/All_Feedback/second_draft/
   19  evince
   20  cd Papers/Hopkins/All_Feedback/second_draft
   21  cd Papers/Hopkins/All_Feedback/second_draft/
   22  evin
   23  history
   24  ls -la sfh_vs_fbmodel.pdf
   25  grep genzel ms.bib
   26  cd
   27  ls -la
   28  mkdir CTA200_2022
   29  ls -la
   30  cd CTA200_2022
   31  cd CTA200_2022/
   32  ls -la
   33  cd
   34  cd CTA200_2022/
   35  ls -la
   36  ls -la > list_dir
   37  ls -la
   38  ci
   39  vim
   40  vm
   41  vi
   42  /pwd
   43  ls -la
   44  cd
   45  ls -la
   46  /mnt/scratch-lustre/student19
   47  cd /mnt/scratch-lustre/student19
   48  mkdir CTA200_2022
   49  cd CTA200_2022
   50  mkdir assignment_1
   51  vim
   52  vim Readme.md
   53  ls -la
   54  cd ~murray
   55  ls -la
   56  find "*Hello*"
   57  find 'Hello'
   58  find Hello
   59  find *hello*
   60  find ~name Hello
   61  find ~ -name Hello
   62  find ~ -name '*Hello'
   63  ls -la
   64  find ~ -name '*Hello'
   65  find ~ -name '*Hello_world'
   66  find/
   67  find/ Hello_world
   68  history
   69  cd
   70  ls -la
   71  cd /mnt/scratch-lustre/student19
   72  ls -la
   73  cd CTA200_2022
   74  ls -la
   75  cd ~murray
   76  ls -la
   77  find 'Hello_World
   78  Hello_world
   79  ls -la
   80  cd
   81  cd /nmt/scratch-lustre/student19
   82  cd ~murray
   83  ls -la
   84  find . -name "Hello*> dev/null
   85  find ~ -name "Hello**"
   86  cd ~murray
   87  ls -la
   88  find . -name "Hello_*2> dev/null
   89  cp
   90  cd ~murray
   91  find . -name "Hello_*2> dev/null
   92  cd ~murray
   93  find . -name "Hello_*2"> dev/null
   94  find . -name "Hello_*2"> /dev/null
   95  grep find . -name "Hello_*2"> /dev/null
   96  grep | find . -name "Hello_*2"> dev/null
   97  ls -la
   98  grep | find . -name "Hello_world*"
   99  cp /Papers/tmp/Hello_world.exe /scratch-lustre/student19/CTA200_2022/assignment_1
  100  cp //Papers/tmp/Hello_world.exe /scratch-lustre/student19/CTA200_2022/assignment_1
  101  cp ~murray/Papers/tmp/Hello_world.exe /scratch-lustre/student19/CTA_2022/assignment_1
  102  cp ~murray/Papers/tmp/Hello_world.exe /mnt/scratch-lustre/student19/CTA_2022/assignment_1
  103  ~murray/Papers/tmp/Hello_world.exe ~student19/mnt/scratch-lustre/student19/CTA_2022/assignment_1
  104  ~murray/Papers/tmp/Hello_world.c ~student19/mnt/scratch-lustre/student19/CTA_2022/assignment_1
  105  cp ~murray/Papers/tmp/Hello_world.c ~student19/mnt/scratch-lustre/student19/CTA_2022/assignment_1
  106  pwd
  107  cd /mnt/scratch-lustre/student19
  108  ls -la
  109  cd CTA200_2022/
  110  ls-la
  111  ls -la
  112  cd assignment_1/
  113  cp ~murray/Papers/tmp/Hello_world.c .
  114  ls -la
  115  cp ~murray/Papers/tmp/Hello_world.py .
  116  cp ~murray/Papers/tmp/Hello_world.exe . 
  117  ls -la
  118  vim
  119  vim Hello_world.exe
  120  mv Hello_world.exe Hello_world_Xu.exe
  121  mv Hello_world.c Hello_world_Xu.c
  122  ls -la
  123  gcc
  124  gcc Hello_world_Xu.c
  125  gcc - v
  126  gcc -v
  127  ls -la
  128  mv a.out Hellow_world.x
  129  ls -la
  130  Hello_world_Xu.c
  131  abc/Hello_world_Xu.exe
  132  $ ./ello_world_Xu.c
  133  $ ./hello_world_Xu.c
  134  $ ./Hello_world_Xu.c
  135  $ ./Hellow_world.x
  136  $ ./Hellow_world
  137  pwd
  138  rm Hello_world_Xu.exe
  139  ls -la
  140  cd ~murray
  141  cd
  142  pwd
  143  cd /mnt/scratch-lustre/student19/CTA200_2022/assignment_1
  144  pwd
  145  ~murray/Papers/tmp/Hello_world.py .
  146  cp ~murray/Papers/tmp/Hello_world.py .
  147  cp ~murray/tmp/Hello_world.py .
  148  ls -la
  149  python Hello_world.py
  150  python3 Hello_world.py
  151  $ ./Hellow_world
  152   ./Hellow_world
  153  ./Hellow_world.x
  154  vim Hello_world.py
  155  vim Hello_world_Xu.c
  156  gcc Hello_world_Xu.c
  157  ls -la
  158  mv Hello_world_Xu.c
  159  mv Hello_world.py Hello_world_Xu.py
  160  vim Hello_world_Xu.c
  161  gcc Hello_world_Xu.c
  162  vim Hello_world_Xu.c
  163  cp ~murray/Papers/tmp/Hello_world.c
  164  cp ~murray/Papers/tmp/Hello_world.c .
  165  vim Hello_world.c
  166  vim Hello_world_Xu.c
  167  gcc Hello_world_Xu.c
  168  ls -la
  169  /a.out
  170  ./a.out
  171  python3 Hello_world_Xu.py
  172  python3 Hello_world_Xu.py > output_python.txt
  173  ls -la
  174  rm Hello_world.c
  175  mv a.out Hellow_world_Xu.x
  176  Hello_world_Xu.c > output_c.txt
  177  ./Hellow_world.x > output_c.txt
  178  ls -la
  179  history
  180  history > history.txt


assingment3/A3Q1.py
#System of Ordinary Differential Equations: Lorenz Euations

import numpy as np
import scipy.integrate
import matplotlib.pyplot as plt

"""
The system of ODEs that we will solve is W ≡XYZ(Lornez Equations)
    X(t) = -σ (X - Y)
    Y(t) = rX - Y - XZ
    Z(t) = -bZ + XY
    Constants:σ,r,b
σ = Prandtl number(ratio of kinematic viscosity to thermal diffusivity)
r = Rayleigh number(dependent on vertical temperature difference between top and bottom of atmosphere)
b = Dimensionless length scale

Initial Conditions:
W0 = [0.,1.,0.]
t = 60
σ = 10.
r = 28
b = 8./3
"""
def XYZ(y,t,σ,r,b):
    X, Y, Z = y

    dydt = [-σ*(X-Y),
            r*X - Y - X*Z,
            -b*Z + X*Y]
    return dydt

def XYZT(y,t,σ,r,b,dt):
    X, Y, Z, T = y
    dydt = [-1*σ(X-Y),
            r*X - Y - X*Z-dt*Y,
            -b*Z + X*Y,
            dt*Y]
    return dydt

def solve_XYZmodel(σ,r,b,ti,tf,y0_XYZ,solve_XYZ=True,dt=0.01,y0_XYZT=[]):
    """
    INPUT:
    σ, r, b, ti, tf, y0_XYZT        #see XYZ function
    OPTIONAL INPUT:
    ::boolean:: solve_XYZ           #whether or not to solve XYZ, default True; if False, solve for XYZT system
    dt, y0_XYZT                      #see XYZT function
    """
    t = np.linspace(ti, tf, 3000)
    if solve_XYZ == True:
        sol_XYZ = scipy.integrate.odeint(XYZ, y0_XYZ, t, args=(σ, r, b))
        plt.title('Lorenz Model')
        plt.plot(t, sol_XYZ[:,0], 'y', label='X(t)')
        plt.plot(t, sol_XYZ[:,1], 'r', label='Y(t)')
        plt.plot(t, sol_XYZ[:,2], 'b', label='Z(t)')
        plt.legend(loc='best')
        plt.xlabel('t')
        plt.grid()
        plt.show()
    else:
        #XYZT model (includes time):
        sol_XYZT = sol_XYZ = scipy.integrate.odeint(XYZT, y0_XYZT, t, args=(σ, r, b, dt))
        plt.title('XYZT Model')
        plt.plot(t, sol_XYZT[:,0], 'y', label='X(t)')
        plt.plot(t, sol_XYZT[:,1], 'r', label='Y(t)')
        plt.plot(t, sol_XYZT[:,2], 'b', label='Z(t)')
        plt.plot(t, sol_XYZT[:,3], 'black', label='D(t)')
        plt.legend(loc='best')
        plt.xlabel('t')
        plt.grid()
        plt.show()

def main():
    σ = 10
    r = 28
    b = 8./3 
    


    y0_XYZ = [0.,1.,0.] #Initial Conditions for XYZ model

    w0_XYZ = [0.,1.00000001,0.] #W0` conditions
    
    dt = 0.01
    y0_XYZT = [0.,1.,0.,60] #Initial Conditions for XYZT model
    
    #set of b and g values to plot
    b_set = [0.2,0.5,0.12,0.4]
    g_set = [0.1,0.1,0.07,0.3]
    try:
        assert(len(b_set) == len(g_set)) #check if both lists are same length
    except AssertionError:
        print("ERROR: Lists not the same length")
        return False
    
    #Interval
    ti = 0
    tf = 200
    
    for i in range(0,len(b_set),1):
        solve_XYZmodel(b_set[i],r,g_set[i],ti,tf,y0_XYZ)

    solve_XYZmodel(σ,r,b,ti,tf,y0_XYZ,False,dt,y0_XYZT)
    solve_XYZmodel(σ,r,b,ti,tf,w0_XYZ,False,dt,y0_XYZT)    

if __name__ == "__main__":
    main()

assingment3/A3Q1plainpy.py

import numpy as np
import matplotlib as plt

N = 1000 #global variable; number of "partitions" for the interval [-2,2]

def function(x,y):
    """
    Takes in two numpy arrays that are result of meshgrid.
    Returns a numpy array with points representing the iteration number for divergence
    """
    max_iter = 100 #maximum number of interations
    c = x + 1j*y
    z = np.zeros((N,N),dtype=complex)
    r = np.zeros((N,N),dtype=int) #return
    mask = np.full((N,N), True, dtype=bool)
    for i in range(0,max_iter,1):
        z[mask] = z[mask]**2 + c[mask]  #z_i = z_i-1**2 + c
        r[mask] = i #i is the iteration number at which point escapes (diverges)
        #if point ever becomes larger than 2, the sequence will escape to infinity:
        mask[np.abs(z) > 2] = False #points that diverge
    return r, mask

def plot_set(bounds,plot_binary):
    """
    Plots the Mandelbrot set with the given bounds
    INPUT:
    ::array:: bounds            #bounds of the plot
    ::boolean:: plot_binary     #whether or not to plot binary (for colour plot, send False)
    """
    x = np.linspace(bounds[0], bounds[1], N)
    y = np.linspace(bounds[2], bounds[3], N)
    xv, yv = np.meshgrid(x,y)
    vals, mandelbrot = function(xv, yv)
    vals = np.flip(vals,0) #flip array vertically
    mandelbrot = np.flip(mandelbrot,0)

    font = {'fontname':'Times New Roman'} 
    plt.figure(dpi=100)
    plt.title('Mandelbrot Set')
    plt.xticks(np.arange(0,N+1,N/4),np.arange(bounds[0],bounds[1]+1,(bounds[1]-bounds[0])/4))
    plt.yticks(np.arange(0,N+1,N/4),np.arange(bounds[3],bounds[2]-1,(bounds[2]-bounds[3])/4))
    plt.xlabel('Re',**font)
    plt.ylabel('Im',**font)
    if plot_binary == False:
        plt.imshow(vals, cmap='plasma_r')
        cbar = plt.colorbar()
        cbar.set_label('Number of Iterations for Divergence')
    else:
        plt.imshow(mandelbrot, cmap=plt.cm.gray)
    plt.show()
    return True

def plot_set_alternate(bounds):
    """
    Similar to plot_set() but instead uses matplotlib.pyplot.contourf()
    """
    x = np.linspace(bounds[0], bounds[1], N)
    y = np.linspace(bounds[2], bounds[3], N)
  
    xv, yv = np.meshgrid(x,y)
    vals, mandelbrot = function(xv, yv)
    
    font = {'fontname':'Times New Roman'} 
    plt.figure(dpi=100)
    plt.title('Mandelbrot Set')
    plt.contourf(xv, yv, vals, levels=100, cmap='plasma_r')
    plt.xlabel('Re',**font)
    plt.ylabel('Im',**font)
    plt.axes().set_aspect('equal')
    plt.colorbar(label='Number of Iterations for Divergence')
    plt.show()

    return True

def main():
    bounds1 = [-2,2,-2,2] #[xmin,xmax,ymin,ymax]
    bounds2 = [-2,0,-1,1]
    plot_set_alternate(bounds1)
    plot_set(bounds2, False)
    plot_set(bounds1, True)
    plot_set(bounds2, True)
    return True

if __name__ == "__main__":
    main()


assingment3/A3Q2plainpy.py
#System of Ordinary Differential Equations: Lorenz Euations

import numpy as np
import scipy.integrate
import matplotlib.pyplot as plt

"""
The system of ODEs that we will solve is W ≡XYZ(Lornez Equations)
    X(t) = -σ (X - Y)
    Y(t) = rX - Y - XZ
    Z(t) = -bZ + XY
    Constants:σ,r,b
σ = Prandtl number(ratio of kinematic viscosity to thermal diffusivity)
r = Rayleigh number(dependent on vertical temperature difference between top and bottom of atmosphere)
b = Dimensionless length scale

Initial Conditions:
W0 = [0.,1.,0.]
t = 60
σ = 10.
r = 28
b = 8./3
"""
def XYZ(y,t,σ,r,b):
    X, Y, Z = y

    dydt = [-σ*(X-Y),
            r*X - Y - X*Z,
            -b*Z + X*Y]
    return dydt

def XYZT(y,t,σ,r,b,dt):
    X, Y, Z, T = y
    dydt = [-1*σ(X-Y),
            r*X - Y - X*Z-dt*Y,
            -b*Z + X*Y,
            dt*Y]
    return dydt

def solve_XYZmodel(σ,r,b,ti,tf,y0_XYZ,solve_XYZ=True,dt=0.01,y0_XYZT=[]):
    """
    INPUT:
    σ, r, b, ti, tf, y0_XYZT        #see XYZ function
    OPTIONAL INPUT:
    ::boolean:: solve_XYZ           #whether or not to solve XYZ, default True; if False, solve for XYZT system
    dt, y0_XYZT                      #see XYZT function
    """
    t = np.linspace(ti, tf, 3000)
    if solve_XYZ == True:
        sol_XYZ = scipy.integrate.odeint(XYZ, y0_XYZ, t, args=(σ, r, b))
        plt.title('Lorenz Model')
        plt.plot(t, sol_XYZ[:,0], 'y', label='X(t)')
        plt.plot(t, sol_XYZ[:,1], 'r', label='Y(t)')
        plt.plot(t, sol_XYZ[:,2], 'b', label='Z(t)')
        plt.legend(loc='best')
        plt.xlabel('t')
        plt.grid()
        plt.show()
    else:
        #XYZT model (includes time):
        sol_XYZT = sol_XYZ = scipy.integrate.odeint(XYZT, y0_XYZT, t, args=(σ, r, b, dt))
        plt.title('XYZT Model')
        plt.plot(t, sol_XYZT[:,0], 'y', label='X(t)')
        plt.plot(t, sol_XYZT[:,1], 'r', label='Y(t)')
        plt.plot(t, sol_XYZT[:,2], 'b', label='Z(t)')
        plt.plot(t, sol_XYZT[:,3], 'black', label='D(t)')
        plt.legend(loc='best')
        plt.xlabel('t')
        plt.grid()
        plt.show()

def main():
    σ = 10
    r = 28
    b = 8./3 
    


    y0_XYZ = [0.,1.,0.] #Initial Conditions for XYZ model

    w0_XYZ = [0.,1.00000001,0.] #W0` conditions
    
    dt = 0.01
    y0_XYZT = [0.,1.,0.,60] #Initial Conditions for XYZT model
    
    #set of b and g values to plot
    b_set = [0.2,0.5,0.12,0.4]
    g_set = [0.1,0.1,0.07,0.3]
    try:
        assert(len(b_set) == len(g_set)) #check if both lists are same length
    except AssertionError:
        print("ERROR: Lists not the same length")
        return False
    
    #Interval
    ti = 0
    tf = 200
    
    for i in range(0,len(b_set),1):
        solve_XYZmodel(b_set[i],r,g_set[i],ti,tf,y0_XYZ)

    solve_XYZmodel(σ,r,b,ti,tf,y0_XYZ,False,dt,y0_XYZT)
    solve_XYZmodel(σ,r,b,ti,tf,w0_XYZ,False,dt,y0_XYZT)    

if __name__ == "__main__":
    main()
